x/incentives/keeper/incentives.go

```diff
@@ -0,0 +1,113 @@
+package keeper
+
+import (
+	"fmt"
+
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/tharsis/evmos/x/incentives/types"
+)
+
+// GetAllIncentives - get all registered Incentives
+func (k Keeper) GetAllIncentives(ctx sdk.Context) []types.Incentive {
+	incentives := []types.Incentive{}
+
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixIncentive)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var incentive types.Incentive
+		k.cdc.MustUnmarshal(iterator.Value(), &incentive)
+
+		incentives = append(incentives, incentive)
+	}
+
+	return incentives
+}
+
+// IterateIncentives iterates over all registered `Incentives` and performs a
+// callback.
+func (k Keeper) IterateIncentives(
+	ctx sdk.Context,
+	handlerFn func(incentive types.Incentive) (stop bool),
+) {
+	store := ctx.KVStore(k.storeKey)
+	iterator := sdk.KVStorePrefixIterator(store, types.KeyPrefixIncentive)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var incentive types.Incentive
+		k.cdc.MustUnmarshal(iterator.Value(), &incentive)
+
+		if handlerFn(incentive) {
+			break
+		}
+	}
+}
+
+// GetIncentive - get registered incentive from the identifier
+func (k Keeper) GetIncentive(
+	ctx sdk.Context,
+	contract common.Address,
+) (types.Incentive, bool) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixIncentive)
+	var incentive types.Incentive
+	bz := store.Get(contract.Bytes())
+	if len(bz) == 0 {
+		return types.Incentive{}, false
+	}
+
+	k.cdc.MustUnmarshal(bz, &incentive)
+	return incentive, true
+}
+
+// SetIncentive stores an incentive
+func (k Keeper) SetIncentive(ctx sdk.Context, incentive types.Incentive) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixIncentive)
+	key := common.HexToAddress(incentive.Contract)
+	bz := k.cdc.MustMarshal(&incentive)
+	store.Set(key.Bytes(), bz)
+}
+
+// DeleteIncentive removes an incentive.
+func (k Keeper) DeleteIncentiveAndUpdateAllocationMeters(ctx sdk.Context, incentive types.Incentive) {
+	store := prefix.NewStore(ctx.KVStore(k.storeKey), types.KeyPrefixIncentive)
+	key := common.HexToAddress(incentive.Contract)
+	store.Delete(key.Bytes())
+
+	// Subtract allocations from allocation meters
+	for _, al := range incentive.Allocations {
+		am, found := k.GetAllocationMeter(ctx, al.Denom)
+		if !found {
+			panic(fmt.Errorf("unable to get allocation meter for denom: %s", al.Denom))
```

Comment:
let's double-check that there's no way this would panic. Otherwise feel free to remove the `found`

Reply:
The incentive is always checked before calling the `DeleteIncentiveAndUpdateAllocationMeters` function. The allocation meters are always updated when an incentive is created or updated/deleted. So removing should be fine.